File: Turquoise/UI/Overlay/TQOverlaySlidingMenu.m
#import "TQOverlaySlidingMenu.h"

@interface TQOverlaySlidingMenu() <UITableViewDataSource, UITableViewDelegate>
@end

@implementation TQOverlaySlidingMenu {
  TQOverlaySlidingMenuPosition _menuPosition;
  UIView *_scrollingView;
  NSString *_title;
  NSArray<NSString *> *_texts;
  NSArray<TQSlidingMenuCallback> *_callbacks;
}

+ (void)showSlidingMenuWithPosition:(TQOverlaySlidingMenuPosition)position
                     verticalOffset:(CGFloat)verticalOffset
                              title:(NSString *)title
                              texts:(NSArray<NSString *> *)texts
                          callbacks:(NSArray<TQSlidingMenuCallback> *)callbacks {
  if (texts.count == 0 || callbacks.count == 0 || callbacks.count < texts.count) {
    return;
  }

  TQOverlaySlidingMenu *menu = [[TQOverlaySlidingMenu alloc] init];
  if (!menu) {
    return;
  }

  menu->_menuPosition = position;
  menu->_title = [title copy];
  menu->_texts = [texts copy];
  menu->_callbacks = [callbacks copy];

  TQOverlay *overlay = [TQOverlay sharedInstance];
  [overlay showWithView:nil animated:NO];
  overlay.slidingMenu = menu;
  overlay.overlayView.manualLayout = YES;

  const CGFloat menuWidth = 200;
  CGFloat menuOriginX = [menu calculateOriginXWithPosition:menu->_menuPosition
                                                     width:menuWidth
                                                    hidden:YES];
  CGFloat menuOriginY = verticalOffset;

  CGFloat maxMenuHeight = CGRectGetMaxY(overlay.overlayView.bounds) - menuOriginY;
  const CGFloat kMenuCellHeight = 45;
  CGFloat menuHeight = kMenuCellHeight * menu->_texts.count;
  if (menu->_title) {
    menuHeight += kMenuCellHeight;
  }

  BOOL tableShouldScroll = NO;
  if (menuHeight > maxMenuHeight) {
    menuHeight = maxMenuHeight;
    tableShouldScroll = YES;
  }

  menu->_scrollingView = [[UIView alloc] initWithFrame:CGRectMake(menuOriginX,
                                                                  menuOriginY,
                                                                  menuWidth,
                                                                  menuHeight)];
  menu->_scrollingView.backgroundColor = [UIColor orangeColor];

  UITableView *tableView = [[UITableView alloc] initWithFrame:menu->_scrollingView.bounds];
  [tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:@"SettingsSliderCell"];
  tableView.backgroundColor = [UIColor blackColor];
  tableView.separatorColor = [UIColor colorWithWhite:.2 alpha:1];
  tableView.separatorInset = UIEdgeInsetsZero;
  tableView.layoutMargins = UIEdgeInsetsZero;
  tableView.dataSource = menu;
  tableView.delegate = menu;
  tableView.scrollEnabled = tableShouldScroll;
  [menu->_scrollingView addSubview:tableView];

  [overlay.overlayView addSubview:menu->_scrollingView];
  [UIView animateWithDuration:[[overlay class] animationDuration]
                   animations:^{
                     CGRect newFrame = menu->_scrollingView.frame;
                     newFrame.origin.x = [menu calculateOriginXWithPosition:menu->_menuPosition
                                                                      width:menuWidth
                                                                     hidden:NO];
                     menu->_scrollingView.frame = newFrame;
                   }];
}

+ (void)dismissSlidingMenuCompletion:(void (^)(BOOL))completion {
  TQOverlay *overlay = [TQOverlay sharedInstance];
  TQOverlaySlidingMenu *slidingMenu = overlay.slidingMenu;
  if (!overlay || !slidingMenu) {
    return;
  }

  [UIView animateWithDuration:[[overlay class] animationDuration]
                   animations:^{
                     CGRect scrollingViewFrame = slidingMenu->_scrollingView.frame;
                     scrollingViewFrame.origin.x =
                         [slidingMenu calculateOriginXWithPosition:slidingMenu->_menuPosition
                                                             width:CGRectGetWidth(slidingMenu->_scrollingView.frame)
                                                            hidden:YES];
                     slidingMenu->_scrollingView.frame = scrollingViewFrame;
                   }
                   completion:^(BOOL finished) {
                     slidingMenu->_texts = nil;
                     slidingMenu->_callbacks = nil;
                     [slidingMenu->_scrollingView removeFromSuperview];
                     overlay.slidingMenu = nil;
                     [overlay dismissAnimated:NO];
                     if (completion) {
                       completion(finished);
                     }
                   }];
}

- (CGFloat)calculateOriginXWithPosition:(TQOverlaySlidingMenuPosition)position
                                  width:(CGFloat)width
                                 hidden:(BOOL)hidden {
  TQOverlay *overlay = [TQOverlay sharedInstance];
  CGRect overlayBounds = overlay.overlayView.bounds;

  switch (position) {
    case TQOverlaySlidingMenuPositionRight:
      return (hidden ? CGRectGetMaxX(overlayBounds) : CGRectGetMaxX(overlayBounds) - width);
    case TQOverlaySlidingMenuPositionLeft:
      return (hidden ? CGRectGetMinX(overlayBounds) - width :  CGRectGetMinX(overlayBounds));
  }
}

- (NSUInteger)groupIndexForIndexPath:(NSIndexPath *)indexPath {
  if (_title) {
    return indexPath.row - 1;
  } else {
    return indexPath.row;
  }
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  NSUInteger numRows = _texts.count;
  if (_title) {
    numRows++;
  }
  return numRows;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"SettingsSliderCell"
                                                          forIndexPath:indexPath];

  cell.textLabel.numberOfLines = 2;
  cell.textLabel.backgroundColor = [UIColor clearColor];
  cell.textLabel.textColor = [UIColor whiteColor];
  cell.textLabel.font = [UIFont fontWithName:@"dungeon" size:12];

  BOOL isTitleCell = _title && [indexPath isEqual:[NSIndexPath indexPathForRow:0 inSection:0]];
  if (isTitleCell) {
    cell.textLabel.font = [UIFont fontWithName:@"dungeon" size:13];
    cell.textLabel.text = _title;
    cell.contentView.backgroundColor = [UIColor blackColor];
    cell.userInteractionEnabled = NO;
  } else {
    NSUInteger textIndex = [self groupIndexForIndexPath:indexPath];
    cell.textLabel.text = _texts[textIndex];
    cell.contentView.backgroundColor = (textIndex % 2 == 0)
        ? [UIColor colorWithRed:0 green:0 blue:.5 alpha:1]
        : [UIColor colorWithRed:0 green:0 blue:1 alpha:1];
    cell.userInteractionEnabled = YES;
  }
  return cell;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  [tableView deselectRowAtIndexPath:indexPath animated:NO];
  NSUInteger groupIndex = [self groupIndexForIndexPath:indexPath];
  TQSlidingMenuCallback callback = [_callbacks[groupIndex] copy];
  [[self class] dismissSlidingMenuCompletion:^(BOOL finished) {
    callback();
  }];
}

@end

File: Turquoise/UI/Overlay/TQOverlaySlidingMenu.swift
import Foundation
import UIKit

public typealias TQSlidingMenuCallback = () -> Void

//typedef void(^TQSlidingMenuCallback)();

@objc public enum TQOverlaySlidingMenuPosition : Int {
  case right
  case left
}

@objc public class TQOverlaySlidingMenu : NSObject, UITableViewDataSource, UITableViewDelegate {
  var menuPosition = TQOverlaySlidingMenuPosition.right
  var scrollingView: UIView?
  var title: String?
  var texts: [String] = []
  var callbacks: [TQSlidingMenuCallback] = []

  // TODO: "callbacks" is currently borked, because the Obj-C<->Swift bridging cannot figure out the blocks in the Obj-C
  //       code correspond to the closures (TQSlidingMenuCallback) here. This will be fixed in an upcoming commit.
  public class func showSlidingMenu(position: TQOverlaySlidingMenuPosition,
                             verticalOffset: CGFloat,
                             title: String,
                             texts: [String],
                             callbacks: Array<Any>) {
    guard !texts.isEmpty && !callbacks.isEmpty && callbacks.count >= texts.count else {
      return
    }

    let menu = TQOverlaySlidingMenu()
    menu.menuPosition = position
    menu.title = title
    menu.texts = texts
    // TODO: fix
//    menu.callbacks = callbacks as! [TQSlidingMenuCallback]

    let overlay = TQOverlay.sharedInstance
    overlay.show(with: nil, animated: false)
    overlay.slidingMenu = menu
    overlay.overlayView?.manualLayout = true

    let menuWidth: CGFloat = 200
    let menuOriginX = menu.calculateOriginX(position: menu.menuPosition,
                                            width: menuWidth,
                                            hidden: true)
    let menuOriginY = verticalOffset


    let maxMenuHeight = (overlay.overlayView?.bounds ?? CGRect.zero).maxY - menuOriginY
    let menuCellHeight: CGFloat = 45.0
    var menuHeight = menuCellHeight * CGFloat(menu.texts.count)
    if menu.title != nil {
      menuHeight += menuCellHeight
    }

    var tableShouldScroll = false
    if menuHeight > maxMenuHeight {
      menuHeight = maxMenuHeight
      tableShouldScroll = true
    }

    menu.scrollingView = UIView(frame: CGRect(x: menuOriginX,
                                              y: menuOriginY,
                                              width: menuWidth,
                                              height: menuHeight))
    menu.scrollingView?.backgroundColor = .orange

    let tableView = UITableView(frame: menu.scrollingView?.bounds ?? .zero)
    tableView.register(UITableViewCell.self, forCellReuseIdentifier: "SettingsSliderCell")
    tableView.backgroundColor = .black
    tableView.separatorColor = UIColor(white: 0.2, alpha: 1)
    tableView.separatorInset = .zero
    tableView.layoutMargins = .zero
    tableView.dataSource = menu
    tableView.delegate = menu
    tableView.isScrollEnabled = tableShouldScroll
    menu.scrollingView?.addSubview(tableView)

    if let scrollingView = menu.scrollingView {
      overlay.overlayView?.addSubview(scrollingView)
    }
    UIView.animate(withDuration: TQOverlay.animationDuration) {
      var newFrame = menu.scrollingView?.frame ?? .zero
      newFrame.origin.x = menu.calculateOriginX(position: menu.menuPosition,
                                                width: menuWidth,
                                                hidden: false)
      menu.scrollingView?.frame = newFrame
    }
  }

  public class func dismissSlidingMenu(completion: ((Bool) -> Void)?) {
    let overlay = TQOverlay.sharedInstance
    guard let slidingMenu = overlay.slidingMenu, let scrollingView = slidingMenu.scrollingView else {
      return
    }

    UIView.animate(withDuration: TQOverlay.animationDuration,
                   animations: {
                    var scrollingViewFrame = scrollingView.frame
                    scrollingViewFrame.origin.x =
                      slidingMenu.calculateOriginX(position: slidingMenu.menuPosition,
                                                   width: scrollingView.frame.width,
                                                   hidden: true)
                    slidingMenu.scrollingView?.frame = scrollingViewFrame
    }) { (finished: Bool) in
      slidingMenu.texts = []
      slidingMenu.callbacks = []
      slidingMenu.scrollingView?.removeFromSuperview()
      overlay.slidingMenu = nil
      overlay.dismiss(animated: false)
      if let completion = completion {
        completion(finished)
      }
    }

  }

  private func calculateOriginX(position: TQOverlaySlidingMenuPosition,
                                width: CGFloat,
                                hidden: Bool) -> CGFloat {
    guard let overlayView = TQOverlay.sharedInstance.overlayView else {
      return 0
    }

    let overlayBounds = overlayView.bounds
    switch position {
    case .right:
      return hidden ? overlayBounds.maxX : overlayBounds.maxX - width

    case .left:
      return hidden ? overlayBounds.minX - width : overlayBounds.minX
    }
  }

  private func groupIndexFor(indexPath: IndexPath) -> Int {
    if self.title != nil {
      return indexPath.row - 1
    } else {
      return indexPath.row
    }
  }

  // MARK: - UITableViewDataSource

  public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    var numRows = self.texts.count
    if self.title != nil {
      numRows += 1
    }
    return numRows
  }

  public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = tableView.dequeueReusableCell(withIdentifier: "SettingsSliderCell", for: indexPath)
    cell.textLabel?.numberOfLines = 2
    cell.textLabel?.backgroundColor = .clear
    cell.textLabel?.textColor = .white
    cell.textLabel?.font = UIFont(name: "dungeon", size: 12.0)

    let isTitleCell = (self.title != nil) && (indexPath == IndexPath(row: 0, section: 0))
    if isTitleCell {
      cell.textLabel?.font = UIFont(name: "dungeon", size: 13.0)
      cell.textLabel?.text = self.title
      cell.contentView.backgroundColor = .black
      cell.isUserInteractionEnabled = false
    } else {
      let textIndex = self.groupIndexFor(indexPath: indexPath)
      cell.textLabel?.text = self.texts[textIndex]
      cell.contentView.backgroundColor = (textIndex % 2 == 0)
        ? UIColor(displayP3Red: 0, green: 0, blue: 0.5, alpha: 1)
        : UIColor(displayP3Red: 0, green: 0, blue: 1.0, alpha: 1)
      cell.isUserInteractionEnabled = true
    }
    return cell
  }

  // MARK: - UITableViewDelegate

  public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    tableView.deselectRow(at: indexPath, animated: false)
    let groupIndex = self.groupIndexFor(indexPath: indexPath)
    // TODO: fix
//    let callback = self.callbacks[groupIndex]
//    TQOverlaySlidingMenu.dismissSlidingMenu { (finished: Bool) in
//      callback()
//    }
  }

}
